File: aoc2025/haskell-solutions/app/Day01.hs
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
import Util

parse :: String -> Int
parse = \case
  ('L':s) -> - read s
  ('R':s) ->   read s

one :: [Int] -> Int
one = length . filter (== 0) . scanl' (\x y -> (x + y) `mod` 100) 50

two :: (Int, Int) -> Int -> (Int, Int)
two (z, d) n = (z + ((d' + abs n) `div` 100), (d + n) `mod` 100)
 where d' = ((signum n * d) `mod` 100)

main :: IO ()
main = do
  inp <- map parse . lines <$> readFile' "../inputs/day01.txt"
  day 1 $ pure (one inp, fst (foldl' two (0,50) inp))

File: aoc2025/clojure-solutions/src/clojure_solutions/day01.clj
(ns clojure-solutions.day01
  (:require [clojure.string :as str]
            [clojure-aoc-util.util :as util]
            [clojure.math :as m])
  (:gen-class))

(defn- parse []
  (->> (slurp "../inputs/day01.txt")
       str/split-lines
       (map (fn [l]
              (read-string
               (str/replace l
                            #"(L)|(R)"
                            (fn [m] (if (= (first m) "L") "-" ""))))))))

(defn- one [xs]
  (->> (reductions (fn [d n] (mod (+ d n) 100)) 50 xs)
       (filter #(= 0 %))
       count))

(defn- two [xs]
  (->> xs
       (reduce
        (fn [[z d] n]
          [(+ z (quot (+ (abs n) (mod (* (m/signum n) d) 100)) 100))
           (mod (+ d n) 100)])
        [0 50])
       first
       int))

(defn -main [& _args]
  (let [inp (parse), f (one inp), s (two inp)]
    (assert (and (= 1011 f) (= 5937 s)))
    (util/print-day 1 f s)))
